cmd/machine-controller-manager/app/controllermanager.go

```diff
@@ -182,32 +182,34 @@ func Run(s *options.MCMServer) error {
 		klog.Fatalf("error creating lock: %v", err)
 	}
 
-	ctx := context.TODO()
 	leaderelection.RunOrDie(ctx, leaderelection.LeaderElectionConfig{
 		Lock:          rl,
 		LeaseDuration: s.LeaderElection.LeaseDuration.Duration,
 		RenewDeadline: s.LeaderElection.RenewDeadline.Duration,
 		RetryPeriod:   s.LeaderElection.RetryPeriod.Duration,
 		Callbacks: leaderelection.LeaderCallbacks{
-			OnStartedLeading: run,
+			OnStartedLeading: startControllers,
 			OnStoppedLeading: func() {
-				klog.Fatalf("leaderelection lost")
+				klog.Errorf("leaderelection lost")
+				waitGroup.Wait()
```

Comment:
okay, but this is for the case where an interrupt is sent to the leader process, which leads to shutdown of workers, but in case when the leader fails to renew lease(due to API latency for example) then there won't be shutdown of the workers right? So we'll keep on waiting?

Reply:
another case which I noticed , is when its attempting to acquire leader lease and we send a `Ctrl-C` then it keeps waiting forever as there are no workers, ideally it should exit with just one interrupt.

Reply:
One solution could be:
```
OnStoppedLeading: func() {
   klog.Errorf("leaderelection lost")
   select{
     case <-ctx.Done():
     default:
       syscall.Kill(syscall.Getpid(), syscall.SIGINT)
   }
   waitGroup.Wait()
}
```

Reply:
Leadership can be lost due to various reasons. Would it not be better if the all controllers that are started when acquiring leadership are subsequently stopped when losing a leadership. There is no need for exiting. `leaderelection.RunOrDie` blocks unless and until its context is cancelled at which point it will return and the `Run` method will complete its job. This would typically happen when that instance is going down in which case there is nothing more to do.

I might have missed something, so please feel free to point that out.